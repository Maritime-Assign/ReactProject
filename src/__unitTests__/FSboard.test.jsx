/**
 * FSBoard.test.jsx
 *
 * This test suite verifies the FSBoard component’s rendering and behavior:
 *
 * 1. Renders "Loading jobs..." before data fetch completes.
 * 2. Displays static table headers after jobs load (no data dependency).
 * 3. Correctly shows both "Open" and "Filled" job states.
 * 4. Handles API failure gracefully (shows error message).
 * 5. Handles empty job list (shows fallback message).
 * 6. Removes loading message after data is fetched.
 * 7. Matches a snapshot to detect unexpected UI changes.
 * 8. Expands and collapses the Notes section when button is clicked.
 */

beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'scrollWidth', {
        configurable: true,
        get() {
            return 200
        },
    })
    Object.defineProperty(HTMLElement.prototype, 'clientWidth', {
        configurable: true,
        get() {
            return 100
        },
    })

    vi.setSystemTime(new Date('2025-01-01'))

    // ✅ Stable random and UUID values (optional but good practice)
    vi.spyOn(Math, 'random').mockReturnValue(0.5)

    // ✅ Flag test env for component logic (used by JobListing)
    window.__vitest_environment__ = true
})

afterAll(() => {
    vi.useRealTimers()
    vi.restoreAllMocks()
})

import {
    render,
    screen,
    waitForElementToBeRemoved,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { expect, vi } from 'vitest'
import '@testing-library/jest-dom'
import { MemoryRouter } from 'react-router-dom'
import FSBoard from '../pages/FSboard'
import getJobsArray from '../components/jobDataAPI'

vi.mock('../auth/AuthContext', () => ({
    UserAuth: () => ({
        user: { id: '1', email: 'test@example.com' },
    }),
}))

vi.mock('../components/jobDataAPI', () => ({
    default: vi.fn(),
}))

describe('FSBoard components', () => {
    beforeEach(() => {
        vi.clearAllMocks()
    })

    test('renders loading message before fetching data', () => {
        render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )
        expect(screen.getByText('Loading jobs...')).toBeInTheDocument()
    })

    test('renders table headers after loading', async () => {
        getJobsArray.mockResolvedValueOnce([{ id: 1 }])
        render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )

        await waitForElementToBeRemoved(() =>
            screen.queryByText('Loading jobs...')
        )

        const headers = [
            'Status',
            'Hall',
            'Date Called',
            'Ship Name',
            'Join Date',
            'Billet',
            'Type',
            'Days',
            'Location',
            'Company',
            'Crew Relieved',
            'Notes',
        ]
        headers.forEach((header) => {
            expect(screen.getByText(header)).toBeInTheDocument()
        })
    })

    test('renders all possible job claim states', async () => {
        getJobsArray.mockResolvedValueOnce([
            { id: 1, open: true, FillDate: null },
            { id: 2, open: false, FillDate: '2025-02-10' },
        ])

        render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )

        await waitForElementToBeRemoved(() =>
            screen.queryByText('Loading jobs...')
        )

        expect(screen.getAllByText('Open')).toHaveLength(1)
        expect(
            screen.getByText(
                (content) =>
                    content.includes('Filled') &&
                    /\d{1,2}\/\d{1,2}\/\d{4}/.test(content)
            )
        ).toBeInTheDocument()
    })

    test('shows error message when data fetch fails', async () => {
        getJobsArray.mockRejectedValueOnce(new Error('Fetch failed'))

        render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )

        await waitForElementToBeRemoved(() =>
            screen.queryByText('Loading jobs...')
        )
        expect(screen.getByText(/error/i)).toBeInTheDocument()
    })

    test('shows fallback UI when no jobs are returned', async () => {
        getJobsArray.mockResolvedValueOnce([])

        render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )

        await waitForElementToBeRemoved(() =>
            screen.queryByText('Loading jobs...')
        )
        expect(screen.getByText(/no jobs/i)).toBeInTheDocument()
    })

    test('removes loading message after successful load', async () => {
        getJobsArray.mockResolvedValueOnce([{ id: 1, open: true }])

        render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )

        await waitForElementToBeRemoved(() =>
            screen.queryByText('Loading jobs...')
        )
        expect(screen.queryByText('Loading jobs...')).not.toBeInTheDocument()
    })

    test('matches snapshot after jobs load', async () => {
        getJobsArray.mockResolvedValueOnce([{ id: 1, open: true, hall: 'OAK' }])

        const { container } = render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )

        await waitForElementToBeRemoved(() =>
            screen.queryByText('Loading jobs...')
        )

        // ✅ Snapshot now stable due to fixed system time and mocks
        expect(container).toMatchSnapshot()
    })

    test('expands and collapses the Notes section when button is clicked', async () => {
        getJobsArray.mockResolvedValueOnce([
            { id: 1, open: true, notes: 'Example job notes for testing' },
        ])

        render(
            <MemoryRouter>
                <FSBoard />
            </MemoryRouter>
        )

        await waitForElementToBeRemoved(() =>
            screen.queryByText('Loading jobs...')
        )

        const allNoteSpans = screen.getAllByText(/example job notes/i)
        const dynamicNoteSpan = allNoteSpans[1] || allNoteSpans[0]

        expect(dynamicNoteSpan).toBeInTheDocument()
        expect(dynamicNoteSpan).toHaveClass('text-ellipsis')

        const expandButton = await screen.findByRole('button', {
            name: /expand notes/i,
        })

        await userEvent.click(expandButton)
        expect(dynamicNoteSpan.className).not.toContain('text-ellipsis')

        await userEvent.click(expandButton)
        expect(dynamicNoteSpan.className).toContain('text-ellipsis')
    })
})
