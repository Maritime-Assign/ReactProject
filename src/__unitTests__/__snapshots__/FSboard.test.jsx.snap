// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FSBoard components > matches snapshot after jobs load 1`] = `
<div>
  <div
    class="w-full flex flex-col my-4 shadow-[0_0_5px_5px_rgba(0,0,0,0.05)] rounded-md"
  >
    <div
      class="grid grid-cols-27 w-full text-center font-bold text-mebablue-dark text-[8px] md:text-sm border-b border-mebagold"
    >
      <div
        class="col-span-2 py-1 font-mont"
      >
        <span>
          Status
        </span>
      </div>
      <div
        class="col-span-1 py-1 font-mont"
      >
        <span>
          Region
        </span>
      </div>
      <div
        class="col-span-1 py-1 font-mont"
      >
        <span>
          Hall
        </span>
      </div>
      <div
        class="col-span-2 py-1 font-mont"
      >
        <span>
          Date Called
        </span>
      </div>
      <div
        class="col-span-3 py-1 font-mont"
      >
        <span>
          Ship Name
        </span>
      </div>
      <div
        class="col-span-2 py-1 font-mont"
      >
        <span>
          Join Date
        </span>
      </div>
      <div
        class="col-span-1 py-1 font-mont"
      >
        <span>
          Billet
        </span>
      </div>
      <div
        class="col-span-2 py-1 font-mont"
      >
        <span>
          Type
        </span>
      </div>
      <div
        class="col-span-1 py-1 font-mont"
      >
        <span>
          Days
        </span>
      </div>
      <div
        class="col-span-3 py-1 font-mont"
      >
        <span>
          Location
        </span>
      </div>
      <div
        class="col-span-2 py-1 font-mont"
      >
        <span>
          Company
        </span>
      </div>
      <div
        class="col-span-3 py-1 font-mont"
      >
        <span>
          Crew Relieved
        </span>
      </div>
      <div
        class="col-span-4 py-1 font-mont"
      >
        <span>
          Notes
        </span>
      </div>
    </div>
    <div
      class="grid grid-cols-27 w-full text-[8px] md:text-[0.8125rem] font-mont font-semibold border-slate-300 border-b overflow-visible h-12 md:h-14.5"
    >
      <div
        class="col-span-2 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <button
          class="inline-flex items-center justify-center px-1 md:px-3 py-1 rounded bg-gradient-to-r from-green-500 to-green-600
                                     text-white hover:from-green-600 hover:to-green-700 disabled:opacity-50 disabled:cursor-not-allowed
                                       transition-all duration-200 ease-out font-medium text-[10px] md:text-sm hover:cursor-pointer"
        >
          Open
        </button>
      </div>
      <div
        class="col-span-1 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-1 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span>
          OAK
        </span>
      </div>
      <div
        class="col-span-2 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-3 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-2 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-1 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-2 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-1 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-3 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-2 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="col-span-3 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span />
      </div>
      <div
        class="relative col-span-4 px-1 items-center justify-center flex h-full bg-slate-200"
      >
        <span
          class="mx-3 p-2 whitespace-nowrap text-ellipsis overflow-hidden transition-all"
          style="display: none; visibility: hidden;"
        />
        <button
          aria-label="Expand Notes"
          class="flex justify-between items-start w-full mx-3 p-2 rounded hover:bg-indigo-200 hover:cursor-pointer transition-all relative"
          style="transition: transform 0.2s, background-color 0.2s, box-shadow 0.2s, border 0.2s;"
          type="button"
        >
          <span
            class="flex-1 overflow-hidden whitespace-nowrap text-ellipsis"
            style="transition: white-space 0.2s;"
          />
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down ml-2 transition-transform duration-200 rotate-0"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
